### PR TITLE
refactor(flow): Unlock flow refactored

### PR DESCRIPTION
**Pull request by amanCypherock**
https://app.clickup.com/t/9002019994/PRF-4859

This PR implements the complete wallet unlock flow with handling of challenge fetching, unlock pow tasks, pin entry and unlock on card operation.

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-4859 Wallet Unlock operation](https://app.clickup.com/t/9002019994/PRF-4859)

File: src/card_operations/card_internal.c [hunk 1]
```diff
@@ -359,7 +359,6 @@ card_error_type_e card_handle_errors(card_operation_data_t *card_data) {
       NFC_RETURN_ABORT_ERROR(card_data, ui_text_card_is_full);
       break;
     case SW_RECORD_NOT_FOUND:
-      card_data->nfc_data.active_cmd_type = WALLET_DOES_NOT_EXISTS_ON_CARD;
```
Comment (ujjwal-cyph) [inline src/card_operations/card_internal.c:362]: why this change?

Comment (amanCypherock) [reply src/card_operations/card_internal.c:362]: yes

File: src/level_four/core/controller/wallet_locked_controller.c [hunk 2]
```diff
@@ -56,123 +55,124 @@
...
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+
+card_error_type_e card_unlock_wallet(const Wallet *wallet) {
+  card_error_type_e result = CARD_OPERATION_DEFAULT_INVALID;
+  card_operation_data_t card_data = {0};
+
+  char heading[50] = "";
   uint8_t wallet_index;
```
[the remark this answers is not in the document]
Comment (amanCypherock) [reply src/level_four/core/controller/wallet_locked_controller.c]: updated

File: src/level_four/core/controller/wallet_locked_controller.c [hunk 3]
```diff
@@ -56,123 +55,124 @@
...
+  get_index_by_name((const char *)wallet->wallet_name, &wallet_index);
```
[the remark this answers is not in the document]
Comment (amanCypherock) [reply src/level_four/core/controller/wallet_locked_controller.c]: updated

File: src/level_four/core/controller/wallet_locked_controller.c [hunk 4]
```diff
@@ -56,123 +55,124 @@
...
+      const char *error_msg = card_data.error_message;
+      if (CARD_OPERATION_SUCCESS == indicate_card_error(error_msg)) {
+        // Re-render the instruction screen
+        instruction_scr_init(ui_text_place_card_below, heading);
+        continue;
       }
-    } else if (tap_card_handle_applet_errors()) {
+    }
+
+    if (CARD_OPERATION_LOCKED_WALLET != card_data.error_type) {
```
[the remark this answers is not in the document]
Comment (amanCypherock) [reply src/level_four/core/controller/wallet_locked_controller.c:170]: updated

**Comment by amanCypherock** [comment]
> A general comment:
> 
> Would it be good to reset challenge in the following cases - `POW_SW_CHALLENGE_FAILED` & `CARD_OPERATION_LOCKED_WALLET`?

For this change I've added zeroising the challenge structure when calling the `set_wallet_locked` API which is already being called in both scenarios.

File: common/interfaces/flash_interface/flash_api.c [hunk 5]
```diff
@@ -728,8 +729,18 @@ int update_wallet_locked_flash(const char *name, const bool is_wallet_locked) {
 
   flash_wallet->is_wallet_locked = is_wallet_locked;
 
-  // Reset nonce to all zero
-  memset(flash_wallet->challenge.nonce, 0, POW_NONCE_SIZE);
+  if (is_wallet_locked) {
```
Comment (ujjwal-cyph) [inline common/interfaces/flash_interface/flash_api.c:732]: add a comment reasoning behind this change.

Comment (amanCypherock) [reply common/interfaces/flash_interface/flash_api.c:739]: updated

**Review by ujjwal-cyph — APPROVED** [review]
